sklearn/metrics/pairwise.py

```diff
@@ -1130,6 +1131,184 @@ def _pairwise_callable(X, Y, metric, **kwds):
                   'russellrao', 'seuclidean', 'sokalmichener',
                   'sokalsneath', 'sqeuclidean', 'yule', "wminkowski"]
 
+DEFAULT_BLOCK_SIZE = 64
+BYTES_PER_FLOAT = 8
+
+
+def _generate_pairwise_distances_blockwise(X, Y=None, metric='euclidean',
+                                           n_jobs=1,
+                                           block_size=DEFAULT_BLOCK_SIZE,
+                                           block_n_rows=1, **kwds):
+    """Generates blocks of the distance matrix from X and optional Y.
+
+    Parameters
+    ----------
+    X : array [n_samples_a, n_samples_a] if metric == "precomputed", or,
+        [n_samples_a, n_features] otherwise
+        Array of pairwise distances between samples, or a feature array.
+
+    Y : array [n_samples_b, n_features], optional
+        An optional second feature array. Only allowed if
+        metric != "precomputed".
+
+    metric : string, or callable
+        The metric to use when calculating distance between instances in a
+        feature array.
+
+    n_jobs : int
+        The number of jobs to use for the computation.
+
+    block_size : int, default=64
+        The maximum number of mebibytes (MiB) of memory per job (see``n_jobs``)
+        to use at a time for calculating pairwise distances.
+
+    block_n_rows : int
+        Number of rows to be computed for each block.
+
+    `**kwds` : optional keyword parameters
+        Any further parameters are passed directly to the distance function.
+
+    Returns
+    -------
+    D : generator of blocks based on the ``block_size`` parameter.
+
+    """
+    if metric != 'precomputed' and Y is None:
+        Y = X
+    n_samples = X.shape[0]
+    for start in range(0, n_samples, block_n_rows):
+        # get distances from block to every other sample
+        stop = min(start + block_n_rows, X.shape[0])
+        yield pairwise_distances(X[start:stop], Y, metric, n_jobs, **kwds)
+
+
+def _generate_pairwise_distances_reduce(X, Y=None, metric='euclidean',
+                                        reduce_func=None, n_jobs=1,
+                                        block_size=DEFAULT_BLOCK_SIZE,
+                                        block_n_rows=1, **kwds):
+    if metric != 'precomputed' and Y is None:
+        Y = X
+    n_samples = X.shape[0]
+    if reduce_func:
+        for start in range(0, n_samples, block_n_rows):
+            # get distances from block to every other sample
+            stop = min(start + block_n_rows, X.shape[0])
+            dist = pairwise_distances(X[start:stop], Y, metric, n_jobs, **kwds)
+            yield reduce_func(dist=dist)
+
+
+def pairwise_distances_reduce(X, Y=None, metric='euclidean', reduce_func=None,
+                              n_jobs=1, block_size=DEFAULT_BLOCK_SIZE, **kwds):
+    if (metric not in _VALID_METRICS and
```

Comment:
I don't get why this function isn't just implemented as `return (reduce_func(D) for D in pairwise_distances_blockwise(...))`